src/helperFunctions/magic.py

```diff
@@ -0,0 +1,43 @@
+"""This is a wrapper around pymagic.
+It aims to provide the same API but with the ability to load multiple magic
+files in the default api.
+"""
+import os
+
+import magic as pymagic
+
+from helperFunctions.fileSystem import get_src_dir
+
+# On ubuntu this is provided by the libmagic-mgc package
+_default_magic = os.getenv('MAGIC', '/usr/lib/file/magic.mgc')
+_fact_magic = f'{get_src_dir()}/bin/firmware.mgc'
+_internal_symlink_magic = f'{get_src_dir()}/bin/internal_symlink.mgc'
+_magic_file = f'{_internal_symlink_magic}:{_fact_magic}:{_default_magic}'
+
+_instances = {}
+
+
+def _get_magic_instance(**kwargs):
+    """Returns an instance of pymagic.Maigc"""
+    # Dicts are not hashable but sorting and creating a tuple is a valid hash
+    key = hash(tuple(sorted(kwargs.items())))
+    i = _instances.get(key)
+    if i is None:
+        i = _instances[key] = pymagic.Magic(**kwargs)
+    return i
+
+
+def from_file(filename, magic_file=_magic_file, **kwargs) -> str:
```

Comment:
also it does not seem to work: I get this output when calling it the first time
```
Warning: using regular magic file `/home/redacted/git/github/FACT_core/src/bin/firmware.mgc'
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 1: Warning: offset `�' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 2: Warning: offset `Firmware v' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 3: Warning: offset `Firmware v' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 5: Warning: offset `VENDOR' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 20: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 21: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 23: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 24: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 25: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 26: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 27: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 28: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 31: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 41: Warning: offset `�' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 45: Warning: offset `�' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 69: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 90: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 336: Warning: offset `
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 342: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 348: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 354: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 360: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 365: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 366: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 367: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 368: Warning: offset `' invalid
/home/redacted/git/github/FACT_core/src/bin/firmware.mgc, 377: Warning: offset `[^d]..|.[^e].|..[^b]|$)' invalid
```

and the mime is not correct (I get `'application/octet-stream'` even though it should be a type that is in the firmware.mgc file)

Also when trying to use the magic file directly with `file` I get the same output and this:
```
file: File 5.45 supports only version 18 magic files. `bin/firmware.mgc' is version 14 (No such file or directory)
```

Reply:
It is actually quite interesting why it works without this changeset.

In fact_helper_file's [setup.py][1] the magic files are compiled using the system's `file`.
Since the package is only ever distributed as sdist and not as a prebuilt wheel on pypi, this ensures that the
magic file version matches the version that the system's libmagic supports.

I think the way to go here is to build multiple versions of firmware.mgc to match all supported versions of libmagic.

[1]: https://github.com/fkie-cad/fact_helper_file/blob/a94a7e0072af468b7891b8b5ecd8366915ceb6f4/setup.py#L32-L43

Reply:
> I think the way to go here is to build multiple versions of firmware.mgc to match all supported versions of libmagic.

Even if you target only the ones that are currently supported in FACT that would still result in five different versions:
- focal: 5.38
- jammy: 5.41
- noble: 5.45
- bullseye: 5.39
- bookworm: 5.44

Maybe compiling on the target system isn't such a bad idea. I'm liking the simplicity of the setup process of `fact_helper_file` more and more where you can simply install the repo as a pip package and directly use the API functions without having to deal with this at all

Reply:
> can just install the repo as a pip package and directly use the API functions without having to deal with this at all

While I agree the pip package is more convenient, I'd argue that the pip package abuses `setup.py` for compiling.
The pip package only works when installing from git and thus compiling the magic database in the process.
The package could not be uploaded to pypi.

I'm not sure what the correct solution is here. I tend to compile the database in FACT's installer.